### PR TITLE
(interval_)sparse_arrayt: Remove unnecessary use of "virtual" and further cleanup

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

**Review by romainbrenguier — COMMENTED** [review]
You will have to update unit tests in solvers/refinement/string_refinement/sparse_array.cpp

The sparse_array class can probably go away and we can only keep the interval one.

**Comment by NathanJPhillips** [comment]
Title says "remove `virtual`" but you actually remove `override`. I'm surprised that an `override` keyword can be removed - it this isn't an override then having it there previously should have been an error.

**Comment by tautschnig** [comment]
We previously had `virtual exprt to_if_expression` and `virtual exprt at`, the former now being `static` and the latter gone completely.

**Comment by tautschnig** [comment]
@romainbrenguier Unit test updated. If you want me to do further cleanup I might need more detailed guidance as I'm not entirely sure replacing each occurrence of `sparse_arrayt` will just work (there is at least one case where an `?:` operation distinguishes between them).

**Review by NathanJPhillips — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 86d057d42caeb0aa9922805f162cfa470e28d58e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78468170

**Review by romainbrenguier — APPROVED** [review]

**Comment by romainbrenguier** [comment]
My assumption was that there was no instance of sparse_arrayt created, but if it's not the case then let's leave it like that.

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -1067,7 +1067,7 @@ static exprt substitute_array_access(
   const bool left_propagate)
 {
   return left_propagate ? interval_sparse_arrayt(expr).to_if_expression(index)
-                        : sparse_arrayt(expr).to_if_expression(index);
+                        : sparse_arrayt::to_if_expression(expr, index);
```
Comment (tautschnig) [inline src/solvers/refinement/string_refinement.cpp:1070]: @romainbrenguier This is the code that does explicitly distinguish `sparse_arrayt` vs `interval_sparse_arrayt`

**Pull request merged**